### PR TITLE
[PORT] Cryotube reagent fix, tweaks and Extinguishes mob on fire

### LINKED ISSUE #3595: Cryotubes not injecting reagents
Cryotubes don't seem to inject reagents consistently. I first noticed this with pyroxadone in a hot tube. 

## Reproduction:

Heat up or cool down your cryotube. 

Add a person in that has damage. Reagents won't inject into the body sometimes. 

A fix to this is to turn off the tube, make the tube the opposite temperature, wait for the temp to be reached then switch the temp back to the original desired temp. 

Deconstructing the cryotube fixes this 100%

<bountysource-plugin>

---
Want to back this issue? **[Post a bounty on it!](https://www.bountysource.com/issues/96476035-cryotubes-not-injecting-reagents?utm_campaign=plugin&utm_content=tracker%2F120077027&utm_medium=issues&utm_source=github)** We accept bounties via [Bountysource](https://www.bountysource.com/?utm_campaign=plugin&utm_content=tracker%2F120077027&utm_medium=issues&utm_source=github).
</bountysource-plugin>

### PR DESCRIPTION
**Pull request by Chessus**
## About The Pull Request
Ports [Makes cryo handle reagent transfer better. #56755](https://github.com/tgstation/tgstation/pull/56755) from TG to fix [#3595](https://github.com/BeeStation/BeeStation-Hornet/issues/3595). It's tweaked. Reagents are now injected every **2 seconds** rather than waiting x amount of time. 

Ports part of [#51092](https://github.com/tgstation/tgstation/pull/51092). Adds the ability for Cryotubes to extinguish a burning corpse when they have been healed. It will eject the occupant when they are fully healed THEN extinguish the person so you can be on fire and cryo-ed no problems.~~Works well with hot cryo~~ on_fire buff doesn't work in the Cryotube.

Pyroxadone metabolism rate is variable now, 0.4u in normal/cold temps and  0.2u in body damaging temperatures. I wanted it to work similar to Cryoxadone but instead gave it a flat rate as cryo works between 3-273k and pyro works above 360.15K with no roof. 

Cryotubes now alert medical when there's no moles of gas to being operation.

Another alert is for when there's no beaker or or reagents are present in the tube and it is turned on. This stops the operation. It will alert medical if either conditions met with a warning. 

Fixed and wrote some more comments.

## Why It's Good For The Game

1. Bug fix, more consistent and faster healing. Less excess chemicals after the operation is finished

2. Burning bodies in Cryotubes is not good as it increases the temperature of the tube, unless it's pyro which is why the extinguishing happens at the end. 

3. This is to balance with the Cryotubes and in general. Speaks for itself. 

4. It didn't before and no one ever analyses the pipes.

5. Stops Cryotubes from working without a beaker/reagents. Warns medical on radio. Also acts as a limiting factor, you shouldn't be in a turned on cryotube (unupgraded) forever

6. Housekeeping 
## Changelog
:cl: Cheesus, ShizCalev, coiax
add: Cryotube extinguishes a burning occupant when they are ejected by the machine. 
code: Comments around cryo cell code
add: Cyrotubes now alerts and stops when the tube has no beaker, reagents or enough moles. 
tweak: changed how much is used and how much is magically transferred to the occupant
tweak: Pyroxadone's metabolization rate is now variable, dependant on heat. 
removed: Outdated cryo reagent injection system
fix: fixed reagents not being injected into an occupant properly.
/:cl:

### CODE REVIEW

**Comment by bloons3** [comment]
Link to PR \#2 is missing

**Comment by Chessus** [comment]
~~There's a lag time in the reagents being injected, it's injects close to the required temperature.~~

File: code/modules/atmospherics/machinery/components/unary_devices/cryo.dm [hunk 1]
```diff
@@ -203,7 +205,7 @@
 			mob_occupant.Unconscious((mob_occupant.bodytemperature * unconscious_factor) * 2000)
 		if(beaker)
 			if(reagent_transfer == 0) // Magically transfer reagents. Because cryo magic.
-				beaker.reagents.trans_to(occupant, 1, efficiency * 0.25) // Transfer reagents.
+				beaker.reagents.trans_to(occupant, 1, efficiency * 25) // Transfer reagents.
```
Comment (park66665) [inline code/modules/atmospherics/machinery/components/unary_devices/cryo.dm]: See: tgstation/tgstation#32565
You should set this to a more modest number, like 1 or 2. (btw that still won't keep up with medicines that gets metabolized quickly; iirc 1~2 is the region where cryoxadone might or might not run out)

Comment (Chessus) [reply code/modules/atmospherics/machinery/components/unary_devices/cryo.dm]: Yeah it seems a bit much. I'm not sure if it's an oversight on op behalf. You're getting like the entire amount of reagents transferred from 1u of of the beaker on lowest efficiency. Not to mention to OD potential.

Comment (Chessus) [reply code/modules/atmospherics/machinery/components/unary_devices/cryo.dm]: @Rukofamicom in your med rework, do you plan to change reagent processing to stop players from loading up on chems?

Comment (Chessus) [reply code/modules/atmospherics/machinery/components/unary_devices/cryo.dm]: @park66665 actually going to use this instead [Makes cryo handle reagent transfer better.](https://github.com/tgstation/tgstation/pull/56755). Removes the reagents_transfer counter but I may need to tweak the maths as our cryoxadone isn't the same as tg's

File: code/modules/atmospherics/machinery/components/unary_devices/cryo.dm [hunk 2]
```diff
@@ -199,16 +202,10 @@
...
-				beaker.reagents.trans_to(occupant, 1, efficiency * 0.25) // Transfer reagents.
-				beaker.reagents.reaction(occupant, VAPOR)
-				air1.adjust_moles(/datum/gas/oxygen, -max(0,air1.get_moles(/datum/gas/oxygen) - 2 / efficiency)) //Let's use gas for this
-			if(++reagent_transfer >= 10 * efficiency) // Throttle reagent transfer (higher efficiency will transfer the same amount but consume less from the beaker).
-				reagent_transfer = 0
-
+			mob_occupant.Sleeping((mob_occupant.bodytemperature * sleep_factor) * 1000 * delta_time)//delta_time is roughly ~2 seconds
+			mob_occupant.Unconscious((mob_occupant.bodytemperature * unconscious_factor) * 1000 * delta_time)
+		if(beaker)//How much to transfer. As efficiency is increased, less reagent from the beaker is used and more is magically transferred to occupant
+			beaker.reagents.trans_to(occupant, (CRYO_TX_QTY / (efficiency * CRYO_MULTIPLY_FACTOR)) * delta_time, efficiency * 15, method = VAPOR)
```
Comment (park66665) [inline code/modules/atmospherics/machinery/components/unary_devices/cryo.dm]:
```suggestion
			beaker.reagents.trans_to(occupant, (CRYO_TX_QTY / (efficiency * CRYO_MULTIPLY_FACTOR)) * delta_time, efficiency * CRYO_MULTIPLY_FACTOR, method = VAPOR)
```

You don't decrease efficiency this way. You keep the calculation *while setting `CRYO_MULTIPLY_FACTOR` to a lower number (e.g. 1~5)

**Review by park66665 — COMMENTED** [review]
I think there was a discussion among people interested in changing the Bee medical code about how low `CRYO_MULTIPLY_FACTOR` should be: i.e. cryo's reagent (magical) multiplication ratio. I thought the consensus was around 1:1 ~ 1:2. I too want the ratio to be more modest, like 1:5. (i.e. `CRYO_MULTIPLY_FACTOR` = 5) We need to discuss this.

File: code/modules/atmospherics/machinery/components/unary_devices/cryo.dm (hunk 2 above)
Comment (Chessus) [reply code/modules/atmospherics/machinery/components/unary_devices/cryo.dm]: I changed it again. I played around with the multiplier equation and settles on what I have now. I also didn't want modify CRYO_MULTIPLY_FACTOR as it's used at the same time and changing it only for the right hand side (multiplier) would effect the left(Amount taken form beaker).

File: code/modules/atmospherics/machinery/components/unary_devices/cryo.dm [hunk 3]
```diff
@@ -1,4 +1,6 @@
 #define CRYOMOBS 'icons/obj/cryo_mobs.dmi'
+#define CRYO_MULTIPLY_FACTOR 25 // Multiply factor is used with efficiency to multiply Tx quantity and how much extra is transfered to occupant magically.
+#define CRYO_TX_QTY 6 // Tx quantity is how much volume should be removed from the cell's beaker - multiplied by delta_time
```
Comment (park66665) [inline code/modules/atmospherics/machinery/components/unary_devices/cryo.dm]:
```suggestion
#define CRYO_MULTIPLY_FACTOR 1.5 // Multiply factor is used with efficiency to multiply Tx quantity and how much extra is transfered to occupant magically.
#define CRYO_TX_QTY 0.6 // Tx quantity is how much volume should be removed from the cell's beaker - multiplied by delta_time
```

File: code/modules/atmospherics/machinery/components/unary_devices/cryo.dm [hunk 4]
```diff
@@ -199,16 +202,10 @@
...
-				beaker.reagents.trans_to(occupant, 1, efficiency * 0.25) // Transfer reagents.
-				beaker.reagents.reaction(occupant, VAPOR)
-				air1.adjust_moles(/datum/gas/oxygen, -max(0,air1.get_moles(/datum/gas/oxygen) - 2 / efficiency)) //Let's use gas for this
-			if(++reagent_transfer >= 10 * efficiency) // Throttle reagent transfer (higher efficiency will transfer the same amount but consume less from the beaker).
-				reagent_transfer = 0
-
+			mob_occupant.Sleeping((mob_occupant.bodytemperature * sleep_factor) * 1000 * delta_time)//delta_time is roughly ~2 seconds
+			mob_occupant.Unconscious((mob_occupant.bodytemperature * unconscious_factor) * 1000 * delta_time)
+		if(beaker)//How much to transfer. As efficiency is increased, less reagent from the beaker is used and more is magically transferred to occupant
+			beaker.reagents.trans_to(occupant, (CRYO_TX_QTY / (efficiency * CRYO_MULTIPLY_FACTOR)) * delta_time, (efficiency * CRYO_MULTIPLY_FACTOR)*(efficiency*0.05), method = VAPOR)
```
Comment (park66665) [inline code/modules/atmospherics/machinery/components/unary_devices/cryo.dm]:
```suggestion
			beaker.reagents.trans_to(occupant, (CRYO_TX_QTY / (efficiency * CRYO_MULTIPLY_FACTOR)) * delta_time, efficiency * CRYO_MULTIPLY_FACTOR, method = VAPOR) // Transfer reagents.
```

**Review by park66665 — CHANGES_REQUESTED** [review]
You *clearly* don't understand the purpose of the defines. If you want different numbers, then you *only* touch the define numbers.

File: code/modules/atmospherics/machinery/components/unary_devices/cryo.dm [hunk 5]
```diff
@@ -175,10 +176,24 @@
...
-	if(!occupant)
+
+	if(!occupant)//Won't operate unless there's an occupant.
+		on = FALSE
+		update_icon()
+		var/msg = "Aborting. No occupant detected."
+		radio.talk_into(src, msg, radio_channel)
+		return
+
+	if(!beaker || !beaker.reagents.reagent_list.len) //No beaker or beaker without reagents with stop the machine from running.
```
Comment (park66665) [inline code/modules/atmospherics/machinery/components/unary_devices/cryo.dm]:
```suggestion
	if(!beaker?.reagents?.reagent_list.len) //No beaker or beaker without reagents with stop the machine from running.
```

File: code/modules/atmospherics/machinery/components/unary_devices/cryo.dm [hunk 6]
```diff
@@ -221,9 +230,11 @@
 
 	var/datum/gas_mixture/air1 = airs[1]
 
-	if(!nodes[1] || !airs[1] || air1.get_moles(/datum/gas/oxygen) < 5) // Turn off if the machine won't work.
+	if(!nodes[1] || !airs[1] || air1.get_moles(/datum/gas/oxygen) < 5) // Turn off if the machine won't work, it will not have enough moles to operate.
```
Comment (park66665) [inline code/modules/atmospherics/machinery/components/unary_devices/cryo.dm]:
```suggestion
	if(!nodes[1] || !airs[1] || air1.get_moles(/datum/gas/oxygen) < 5) // Turn off if the machine won't work due to not having enough moles to operate.
```

File: code/modules/atmospherics/machinery/components/unary_devices/cryo.dm [hunk 7]
```diff
@@ -221,9 +230,11 @@
...
 		on = FALSE
 		update_icon()
+		var/msg = "Aborting. Not enough moles of gas present to operate."
```
Comment (park66665) [inline code/modules/atmospherics/machinery/components/unary_devices/cryo.dm]:
```suggestion
		var/msg = "Aborting. Not enough gas present to operate."
```

File: _maps/_basemap.dm [hunk 8]
```diff
@@ -1,4 +1,4 @@
-//#define LOWMEMORYMODE //uncomment this to load centcom and runtime station and thats it.
+#define LOWMEMORYMODE //uncomment this to load centcom and runtime station and thats it.
```
Comment (park66665) [inline _maps/_basemap.dm]:
```suggestion
//#define LOWMEMORYMODE //uncomment this to load centcom and runtime station and thats it.
```

You generally shouldn't include *this* in your PR.

**Review by park66665 — COMMENTED** [review]
After these are addressed, this PR should be good to go.

File: code/modules/atmospherics/machinery/components/unary_devices/cryo.dm (hunk 5 above)
Comment (Chessus) [reply code/modules/atmospherics/machinery/components/unary_devices/cryo.dm]: Intereasting!

**Comment by Chessus** [comment]
@park66665 The user now gets 0.8u instead of 1.2u. Ratio hasn't changed and this fixed the excess reagents. With 50 brute damage, a naked person ends up with ~6-7u of cryo/pyro when the operation is finished. That's more than enough for them to return to normal body temp.

**Review by utilbird — APPROVED** [review]

**Review by park66665 — APPROVED** [review]
On a side note, this PR is shown to cause an increase in cryoxadone consumption in testing environments. Players and coders are welcome to open a PR if they wish to tweak the cryoxadone consumption rate.

All clear. This PR is good to go.

**Pull request merged**